// @zilliqa-js/util/dist/bytes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * intToHexArray
 *
 * @param {number} int - the number to be converted to hex
 * @param {number)} size - the desired width of the hex value. will pad.
 *
 * @returns {string[]}
 */
exports.intToHexArray = function (int, size) {
    var hex = [];
    var hexRep = [];
    var hexVal = int.toString(16);
    // TODO: this really needs to be refactored.
    for (var i = 0; i < hexVal.length; i++) {
        hexRep[i] = hexVal[i].toString();
    }
    for (var i = 0; i < size - hexVal.length; i++) {
        hex.push('0');
    }
    for (var i = 0; i < hexVal.length; i++) {
        hex.push(hexRep[i]);
    }
    return hex;
};
/**
 * intToByteArray
 *
 * Converts a number to Uint8Array
 *
 * @param {number} num
 * @param {number} size
 *
 * @returns {Uint8Array}
 */
exports.intToByteArray = function (num, size) {
    var x = num;
    var res = [];
    while (x > 0) {
        res.push(x & 255);
        x = x >> 8;
    }
    var pad = size - res.length;
    for (var i = 0; i < pad; i++) {
        res.unshift(0);
    }
    return Uint8Array.from(res);
};
/**
 * hexToByteArray
 *
 * Convers a hex string to a Uint8Array
 *
 * @param {string} hex
 * @returns {Uint8Array}
 */
exports.hexToByteArray = function (hex) {
    var res = new Uint8Array(hex.length / 2);
    for (var i = 0; i < hex.length; i += 2) {
        res[i / 2] = parseInt(hex.substring(i, i + 2), 16);
    }
    return res;
};
/**
 * hexToIntArray
 *
 * @param {string} hex
 * @returns {number[]}
 */
exports.hexToIntArray = function (hex) {
    if (!hex || !exports.isHex(hex)) {
        return [];
    }
    var res = [];
    for (var i = 0; i < hex.length; i++) {
        var c = hex.charCodeAt(i);
        var hi = c >> 8;
        var lo = c & 0xff;
        hi ? res.push(hi, lo) : res.push(lo);
    }
    return res;
};
/**
 * pack
 *
 * Takes two 16-bit integers and combines them. Used to compute version.
 *
 * @param {number} a
 * @param {number} b
 *
 * @returns {number} - a 32-bit number
 */
exports.pack = function (a, b) {
    if (a >> 16 > 0 || b >> 16 > 0) {
        throw new Error('Both a and b must be 16 bits or less');
    }
    return (a << 16) + b;
};
/**
 * compareBytes
 *
 * A constant time HMAC comparison function.
 *
 * @param {string} a
 * @param {string} b
 * @returns {boolean}
 */
exports.isEqual = function (a, b) {
    var bzA = exports.hexToIntArray(a);
    var bzB = exports.hexToIntArray(b);
    if (bzA.length !== bzB.length) {
        return false;
    }
    var result = 0;
    for (var i = 0; i < bzA.length; i++) {
        result |= bzA[i] ^ bzB[i];
    }
    return result === 0;
};
/**
 * isHex
 *
 * @param {string} str - string to be tested
 * @returns {boolean}
 */
exports.isHex = function (str) {
    var plain = str.replace('0x', '');
    return /[0-9a-f]*$/i.test(plain);
};
