// @zilliqa-js/util/dist/validation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var bn_js_1 = tslib_1.__importDefault(require("bn.js"));
var long_1 = tslib_1.__importDefault(require("long"));
exports.isAddress = function (address) {
    return exports.isByteString(address, 40);
};
exports.isPrivateKey = function (privateKey) {
    return exports.isByteString(privateKey, 64);
};
exports.isPubKey = function (pubKey) {
    return exports.isByteString(pubKey, 66);
};
exports.isSignature = function (sig) {
    return exports.isByteString(sig, 128);
};
exports.isByteString = function (str, len) {
    return !!str.replace('0x', '').match("^[0-9a-fA-F]{" + len + "}$");
};
exports.isNumber = function (x) {
    return typeof x === 'number';
};
exports.isBN = function (x) {
    return bn_js_1.default.isBN(x);
};
exports.isLong = function (x) {
    return long_1.default.isLong(x);
};
exports.isString = function (x) {
    return typeof x === 'string';
};
exports.isPlainObject = function (x) {
    if (typeof x === 'object' && x !== null) {
        var proto = Object.getPrototypeOf(x);
        return proto === Object.prototype || proto === null;
    }
    return false;
};
var PRAGMA_REQUIRED = '@@ZJS_REQUIRED@@';
exports.required = function (fn) {
    if (typeof fn === 'function') {
        return Object.defineProperty(fn, 'required', {
            value: PRAGMA_REQUIRED,
        });
    }
    throw new Error('fn is not a function');
};
exports.matchesObject = function (x, test) {
    var e_1, _a;
    if (exports.isPlainObject(x)) {
        for (var key in test) {
            if (test.hasOwnProperty(key)) {
                try {
                    for (var _b = tslib_1.__values(test[key]), _c = _b.next(); !_c.done; _c = _b.next()) {
                        var tester = _c.value;
                        var value = x[key];
                        if (typeof value === 'undefined' && tester.required) {
                            throw new Error('Key not found: ' + key);
                        }
                        else {
                            continue;
                        }
                        if (typeof tester !== 'function') {
                            throw new Error('Validator is not a function');
                        }
                        if (!tester(value)) {
                            throw new Error('Validation failed for ' + key);
                        }
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
            }
        }
    }
    return true;
};
